src/nonstandard.jl

```diff
@@ -56,3 +56,18 @@ constructorof(::Type{<:LinRange}) = linrange_constructor
 ### Expr: args get splatted
 # ::Expr annotation is to make it type-stable on Julia 1.3-
 constructorof(::Type{<:Expr}) = (head, args) -> Expr(head, args...)::Expr
+
+### Cholesky
+setproperties(C::LinearAlgebra.Cholesky, patch::NamedTuple{()}) = C
+function setproperties(C::LinearAlgebra.Cholesky, patch::NamedTuple{(:L,)})
+    return LinearAlgebra.Cholesky(C.uplo === 'U' ? permutedims(patch.L) : patch.L, C.uplo, C.info)
+end
+function setproperties(C::LinearAlgebra.Cholesky, patch::NamedTuple{(:U,)})
+    return LinearAlgebra.Cholesky(C.uplo === 'L' ? permutedims(patch.U) : patch.U, C.uplo, C.info)
+end
```

Comment:
Wouldn't it be more efficient (at least in the first place) to change `C.uplo` instead of permuting `patch.L` or `patch.U`?

Reply:
Is there a common convention in ConstructionBase? I think there's also no guarantee that `permutedims` keeps the type of the array invariant, so there could still be surprises for users (and type instabilities?).

Reply:
@jw3126 maybe knows?

Reply:
Yes we love it if the following invariants hold:
```julia
patch = getproperties(obj) # or a subset of this
@test obj == setproperties(obj, patch)
@test typeof(obj) == typeof(setproperties(obj, patch))
```

Reply:
That would be satisfied in both cases, it seems.

Reply:
Ah, so no type parameter changed if the type of L changes?

Reply:
Wait, but won't `C.UL` change if we change `uplo`?

Reply:
> Ah, so no type parameter changed if the type of L changes?

If you change the type of `L` (or some other property) then the type of the resulting object after calling `setproperties` will be different. But as far as I understood, the invariants you mentioned are only concerned with calling `setproperties` with the values (and types) of the current properties?

Reply:
Oh no, wait a second: Because https://github.com/JuliaLang/julia/blob/a5b2197fe46a2ce2db4be39a6d79132dffd31414/stdlib/LinearAlgebra/src/cholesky.jl#L516-L528 wraps all properties either in `UpperTriangular` or `LowerTriangular`, in the current form all variants will change the type?!

Reply:
I wonder if it's possible at all to satisfy the invariants: even after unwrapping the `Triangular` wrappers we're left only with the output of `copy(obj.A')` in some cases which might be of a different type than `obj.A` (and it's not guaranteed to be "reverted" by `permutedims` or `copy • adjoint`).

Reply:
If it is impossible to keep the type the same, then so be it.

Reply:
I'm confuse; the current implementation doesn't change the type? Changing the `uplo` will, no?

Current:
```julia
julia> C = Cholesky(X, 'L', 0)
Cholesky{Float64, Matrix{Float64}}
L factor:
3×3 LowerTriangular{Float64, Matrix{Float64}}:
 5.98603   ⋅         ⋅ 
 1.81901  1.06351    ⋅ 
 2.4642   0.806017  1.34749

julia> # Update `L`.
       C_new = ConstructionBase.setproperties(C, (L=2 .* parent(C.L),))
Cholesky{Float64, Matrix{Float64}}
L factor:
3×3 LowerTriangular{Float64, Matrix{Float64}}:
 11.9721    ⋅        ⋅ 
  3.63802  2.12701   ⋅ 
  4.9284   1.61203  2.69499

julia> # Update `L`.
       C_new = ConstructionBase.setproperties(C, (U=2 .* parent(C.U),))
Cholesky{Float64, Matrix{Float64}}
L factor:
3×3 LowerTriangular{Float64, Matrix{Float64}}:
 11.9721    ⋅        ⋅ 
  3.63802  2.12701   ⋅ 
  4.9284   1.61203  2.69499

julia> # Update `L`.
       C_new = ConstructionBase.setproperties(C, (U=2 .* C.U,))
Cholesky{Float64, Matrix{Float64}}
L factor:
3×3 LowerTriangular{Float64, Matrix{Float64}}:
 11.9721    ⋅        ⋅ 
  3.63802  2.12701   ⋅ 
  4.9284   1.61203  2.69499
```

Reply:
The type will still be the same in these examples, even if you change `uplo` - it will always be `Cholesky{Float64, Matrix{Float64}}`. It's just the `show` method that makes it look different.

Generally, I think the example is a bit too simple to reason about the problematic cases.

Reply:
Here's what happens if you change the `uplo` instead:
```julia
julia> using ConstructionBase, LinearAlgebra

julia> x = randn(3, 3)
3×3 Matrix{Float64}:
  1.7626    -1.10158   -1.8708
  0.608564   1.64422    0.354848
 -2.28876    0.717459   0.55325

julia> X = x * x'
3×3 Matrix{Float64}:
  7.82013  -1.40245    -5.85952
 -1.40245   3.19974    -0.0168719
 -5.85952  -0.0168719   6.05924

julia> C = Cholesky(X, 'L', 0)
Cholesky{Float64, Matrix{Float64}}
L factor:
3×3 LowerTriangular{Float64, Matrix{Float64}}:
  7.82013    ⋅          ⋅ 
 -1.40245   3.19974     ⋅ 
 -5.85952  -0.0168719  6.05924

julia> C = Cholesky(X, 'L', 0)
Cholesky{Float64, Matrix{Float64}}
L factor:
3×3 LowerTriangular{Float64, Matrix{Float64}}:
  7.82013    ⋅          ⋅ 
 -1.40245   3.19974     ⋅ 
 -5.85952  -0.0168719  6.05924

julia> # Update `L`.
       C_new = ConstructionBase.setproperties(C, (L=2 .* parent(C.L),))
Cholesky{Float64, Matrix{Float64}}
L factor:
3×3 LowerTriangular{Float64, Matrix{Float64}}:
  15.6403     ⋅           ⋅ 
  -2.80489   6.39948      ⋅ 
 -11.719    -0.0337438  12.1185

julia> # Update `U`.
       C_new = ConstructionBase.setproperties(C, (U=2 .* parent(C.U),))
Cholesky{Float64, Matrix{Float64}}
U factor:
3×3 UpperTriangular{Float64, Matrix{Float64}}:
 15.6403  -2.80489  -11.719
   ⋅       6.39948   -0.0337438
   ⋅        ⋅        12.1185

julia> # Update `U`.
       C_new = ConstructionBase.setproperties(C, (U=2 .* C.U,))
Cholesky{Float64, UpperTriangular{Float64, Matrix{Float64}}}
U factor:
3×3 UpperTriangular{Float64, Matrix{Float64}}:
 15.6403  -2.80489  -11.719
   ⋅       6.39948   -0.0337438
   ⋅        ⋅        12.1185
```

The patch:
```
diff --git a/src/nonstandard.jl b/src/nonstandard.jl
index 26f1c18..b11749b 100644
--- a/src/nonstandard.jl
+++ b/src/nonstandard.jl
@@ -60,10 +60,10 @@ constructorof(::Type{<:Expr}) = (head, args) -> Expr(head, args...)::Expr
 ### Cholesky
 setproperties(C::LinearAlgebra.Cholesky, patch::NamedTuple{()}) = C
 function setproperties(C::LinearAlgebra.Cholesky, patch::NamedTuple{(:L,)})
-    return LinearAlgebra.Cholesky(C.uplo === 'U' ? permutedims(patch.L) : patch.L, C.uplo, C.info)
+    return LinearAlgebra.Cholesky(patch.L, 'L', C.info)
 end
 function setproperties(C::LinearAlgebra.Cholesky, patch::NamedTuple{(:U,)})
-    return LinearAlgebra.Cholesky(C.uplo === 'L' ? permutedims(patch.U) : patch.U, C.uplo, C.info)
+    return LinearAlgebra.Cholesky(patch.U, 'U', C.info)
 end
 function setproperties(C::LinearAlgebra.Cholesky, patch::NamedTuple{(:UL,)})
     return LinearAlgebra.Cholesky(patch.UL, C.uplo, C.info)
```

Reply:
The problem is that you don't unwrap the matrix in the proposed `setproperties` - as in the `Cholesky(::LowerTriangular)` etc. constructors in LinearAlgebra you should not call the constructor with `patch.L` but `patch.L.data` if it's a `LowerTriangular`.

Reply:
The `LowerTriangular` etc. only show up in the properties and hence the `show` method in LinearAlgebra, the data is stored unwrapped.

Reply:
Ah, wait, so you're with me on the `uplo` question now, but you're saying we should just add a `parent` in our current impl? I had this at some earlier point, but see I forgot to include it in the current version.

Reply:
Added the type-checking to the tests btw.

Reply:
If you unwrap, the type will always stay `Cholesky{Float64, Matrix{Float64}}` when changing `uplo`. The `show` output changes but the type does not, so the invariants are still satisfied.

Reply:
Yeah yeah, the type stays the same, _but_ the type of the property `UL` changes if you change `uplo`; that seems non-desirable to me, no? :confused:

Reply:
Basically, it's a question of whether we want
```julia
typeof(setproperties(C, patch).UL) === typeof(C.UL)
```
to hold or not.

Reply:
Yes so we want 
```julia
patch = getproperties(obj) 
obj ==== setproperties(obj, patch)
```
where `====` means as equal as possible. At least `isequal` should hold. If fields change types, that is not so nice, but I could live with that in this case.

Reply:
So in both of the scenarios I and @devmotion  describe, the _fields_ don't change types, only the _properties_ do (in particular, the _property_ `C.UL`'s type is basically run-time information dependent on the char-field `C.uplo`)

Reply:
properties changing types is also not very nice, but I can also live with that. What we are after is that exchanging `obj` for  `setproperties(obj, patch)` should not break any reasonable code.

Reply:
Gotcha :+1: 

So do we stick with the current implementation then? This has the property that the type of `C.UL` persists, even if the user tries to update the `U` field when the underlying `Cholesky` is working with lower-triangular matrices (but it results in a `permutedims` call), while what @devmotion does not preserve the type _but_ it has the benefit of avoiding the `permutedims` all together.

No matter, this probably shouldn't hold this PR up for so long :sweat_smile: IMO users shouldn't really be calling `setproperties` with an upper-triangular matrix if the factor is actually working with lower-triangular, and vice versa, so whatever we go with probably doesn't make much of a difference. 

Personally I'm in favour of the additional memory-allocs to avoid silently changing `uplo`, but am of course happy to make the change if both of you prefer the alternative :+1:

Reply:
Isn't it a bit unexpected to convert a user-provided upper-triangular matrix to a lower-triangular matrix?

In any case, similar to the `Cholesky` constructors in LinearAlgebra, I think we should maybe make the code a bit safer by only unwrapping `LowerTriangular` and `UpperTriangular` but not arbitrary user-provided matrices (such as views etc.). The latter seems slightly dangerous as it could lead to surprising and unexpected results.

Reply:
> No matter, this probably shouldn't hold this PR up for so long 😅

Agreed, I am happy to merge whatever you guys think is the best `setproperties` choice for `Cholesky`. I never needed to manipulate `Cholesky` factors like this myself before, so you are in a better spot to make the call anyway.

Reply:
> I never needed to manipulate Cholesky factors like this myself before, so you are in a better spot to make the call anyway.

My feeling is that generally it shouldn't be encouraged to update `Cholesky` factors with `setproperties` - if you have a triangular Cholesky factor you should just call the one-argument `Cholesky` constructor directly.

Reply:
> My feeling is that generally it shouldn't be encouraged to update Cholesky factors with setproperties - if you have a triangular Cholesky factor you should just call the one-argument Cholesky constructor directly.

Don't disagree with this, _but_ `ConstructionBase.setproperties` is used heavily in the ecosystem, e.g. Setfields.jl, and so we should definitively allow this, i.e. implement `ConstructionBase.setproperties`; whether we should encourage it or not is a different question.

Reply:
> Isn't it a bit unexpected to convert a user-provided upper-triangular matrix to a lower-triangular matrix?

Yeah, that is very true, but it seems to me to be less weird than not even setting the property the user requests to be set :man_shrugging: 

>  In any case, similar to the Cholesky constructors in LinearAlgebra, I think we should maybe make the code a bit safer by only unwrapping LowerTriangular and UpperTriangular but not arbitrary user-provided matrices (such as views etc.). The latter seems slightly dangerous as it could lead to surprising and unexpected results.

But this I'm very happy with yeah :+1:

Reply:
I've now made it so it only works with `LowerTriangular` and `UpperTriangular`, which should make things a bit safer:)

Reply:
You know what, I'm happy to change the `uplo` if you prefer @devmotion :+1: 

I'm still slightly in favour of not changing `uplo` but I don't think it matters too much:)

Reply:
Btw, `permutedims` materializes the matrix it seems, but I'm uncertain if this is an intended effect of `permutedims` or if it's just that it hasn't been sufficiently specialized yet :confused: 

So I replaced it with `copy(transpose(...))` on the underlying data, which should be much safter.